### PR TITLE
Add memory usage to contexts

### LINKED ISSUE #1826: Report used memory/max used memory
### Problem Statement

In order to better understand the memory usage, Sentry should report the used memory of an application.

### Solution Brainstorm

Dart offers to query some memory information:
- https://api.dart.dev/stable/3.2.5/dart-io/ProcessInfo/currentRss.html
- https://api.dart.dev/stable/3.2.5/dart-io/ProcessInfo/maxRss.html
- https://api.dart.dev/stable/3.2.5/dart-developer/NativeRuntime/writeHeapSnapshotToFile.html (most likely not available in release mode tho)
- There's also https://github.com/dart-lang/leak_tracker/tree/main/pkgs/memory_usage doing some more

These APIs make it available for Dart and Flutter.

### Are you willing to submit a PR?

Yes

One problem with this is, that this is a pretty new API, meaning using it would break for currently supported lower versions.

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

Add memory usage to contexts

## :bulb: Motivation and Context

Closes #1826

## :green_heart: How did you test it?

Sample app

## :pencil: Checklist

- [x] I reviewed submitted code
- [ ] I added tests to verify changes
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

**Comment by buenaflor** [comment]
We could use these fields if they make sense.

```
memory_size
Optional. Total system memory available in bytes.

free_memory
Optional. Free system memory in bytes.

usable_memory
Optional. Memory usable for the app in bytes.
```

see https://develop.sentry.dev/sdk/event-payloads/contexts/

**Comment by buenaflor** [comment]
app_memory might also be useful

```
app_memory
Optional. Amount of memory used by the application in bytes.
```

depending on the meaning of app_memory both current and max rss could fit here. we'd need more clarification for that

**Comment by buenaflor** [comment]
Okay it's supposed to be the memory at the time of the event so `currentRss` should fit here

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 - Capture total frames, frames delay, slow & frozen frames and attach to spans ([#2106](https://github.com/getsentry/sentry-dart/pull/2106))
 - Support WebAssembly compilation (dart2wasm) ([#2113](https://github.com/getsentry/sentry-dart/pull/2113))
+- Add memory usage to contexts ([#2133](https://github.com/getsentry/sentry-dart/pull/2133))
```
Comment (buenaflor) [inline CHANGELOG.md]: pls specify that web is not included (if there is no way to retrieve memory usage on web)

File: dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart [hunk 2]
```diff
@@ -40,6 +40,11 @@ class IoEnricherEventProcessor implements EnricherEventProcessor {
     );
 
     contexts['dart_context'] = _getDartContext();
+    contexts['process_info'] = <String, dynamic>{
+      'currentResidentSetSize':
+          _bytesToHumanReadableFileSize(ProcessInfo.currentRss),
+      'maxResidentSetSize': _bytesToHumanReadableFileSize(ProcessInfo.maxRss),
```
Comment (buenaflor) [inline dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart]: are there any alternatives for web?

File: dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart [hunk 3]
```diff
@@ -40,6 +40,11 @@ class IoEnricherEventProcessor implements EnricherEventProcessor {
     );
 
     contexts['dart_context'] = _getDartContext();
+    contexts['process_info'] = <String, dynamic>{
+      'currentResidentSetSize':
```
Comment (buenaflor) [inline dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart]: let's add currentRss to app_memory

**Review by buenaflor — COMMENTED** [review]
left some comments,

I'm not sure though how we should categorize maxRss

File: dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart (hunk 2 above)
Comment (denrase) [reply dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart]: Found this, but there's not much documentation: https://api.dart.dev/stable/2.18.0/dart-html/MemoryInfo/usedJSHeapSize.html

File: dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart (hunk 3 above)
Comment (ueman) [reply dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart]: How does `currentRss` compare to what the native side sets? Does it override what the native side sets?

File: dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart (hunk 2 above)
Comment (buenaflor) [reply dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart]: this is based on this https://developer.mozilla.org/en-US/docs/Web/API/Performance/memory

it's deprecated and not all browsers support it, dunno if it's worth the effort to add this if we have to remove it again at some point

Comment (denrase) [reply dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart]: Yeah, i removed it after reading the docs and added some info.

File: dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart (hunk 3 above)
Comment (denrase) [reply dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart]: Updated the code and we only set these values if there are no native SDKs

Comment (buenaflor) [reply dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart]: I don't think android sets it, I know iOS for sure does @denrase

**Comment by denrase** [comment]
@buenaflor Used the mapping below, as these seemed the most appropriate accodring to the documentation. WDYT?

Used

- app_memory -> **currentRss**
Optional. Amount of memory used by the application in bytes.

- usable_memory -> **maxRss**
Optional. Memory usable for the app in bytes.

Values below are not used, as we can't say anything about the system memory as far as I can see.

- memory_size
Optional. Total system memory available in bytes.

- free_memory
Optional. Free system memory in bytes.

**Comment by denrase** [comment]
There's a package to get more system info we could look into: https://pub.dev/packages/system_info2

**Comment by denrase** [comment]
Here's how it checks for physical memory info on different platforms.

https://github.com/onepub-dev/system_info/blob/8a9bf6b8eb7c86a09b3c3df4bf6d7fa5a6b50732/lib/src/platform/memory.dart#L73

**Comment by buenaflor** [comment]
> usable_memory -> maxRss
Optional. Memory usable for the app in bytes.

hm not sure if this completely fits since maxRss captures the highest amount used thus far

**Comment by buenaflor** [comment]
> There's a package to get more system info we could look into: https://pub.dev/packages/system_info2

can those commands be run in production?

File: dart/lib/src/event_processor/enricher/io_platform_memory.dart [hunk 4]
```diff
@@ -0,0 +1,107 @@
+import 'dart:io';
+
+// Get total & free platform memory (in bytes) for linux and windows operating systems.
+// Source: https://github.com/onepub-dev/system_info/blob/8a9bf6b8eb7c86a09b3c3df4bf6d7fa5a6b50732/lib/src/platform/memory.dart
+class PlatformMemory {
+  PlatformMemory(this.operatingSystem);
```
Comment (ueman) [inline dart/lib/src/event_processor/enricher/io_platform_memory.dart]: You could pass a `Platform` instead and use that to check the platform.
E.g.
```dart
if (platform.isLinux) {
  // ...
} else if (platform.isWindows) {
  // ...
}
```

File: dart/lib/src/event_processor/enricher/io_platform_memory.dart [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  String? _exec(String executable, List<String> arguments,
+      {bool runInShell = false}) {
+    try {
+      final result =
+          Process.runSync(executable, arguments, runInShell: runInShell);
+      if (result.exitCode == 0) {
+        return result.stdout.toString();
+      }
+    } catch (e) {
+      //
```
Comment (buenaflor) [inline dart/lib/src/event_processor/enricher/io_platform_memory.dart]: maybe we could log here with `options.logger` so exceptions are not swallowed

File: CHANGELOG.md [hunk 6]
```diff
@@ -6,6 +6,9 @@
...
+  - Only for Linux/Windows applications, as iOS/Android use native SDKs
```
Comment (buenaflor) [inline CHANGELOG.md]: and also mention macOS

**Review by buenaflor — COMMENTED** [review]
left some comments, regarding maxRss we can also name it ourselves, not sure if usable_memory is the best fit. wdyt?

File: dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart (hunk 3 above)
Comment (denrase) [reply dart/lib/src/event_processor/enricher/io_enricher_event_processor.dart]: Updated so we set `appMemory` to `ProcessInfo.currentRss`, it it's not already set, regardless of native integration. The other values are merged, so this is additional information.

On Linux/Windows we now also add free and total system memory, if possible.

**Comment by denrase** [comment]
> can those commands be run in production?

Testes in debug in am Windows VM and tests pass so far.

File: CHANGELOG.md [hunk 7]
```diff
@@ -17,6 +17,8 @@
   - Use `SentryFlutter.pauseAppHangTracking()` and `SentryFlutter.resumeAppHangTracking()`
 - Capture total frames, frames delay, slow & frozen frames and attach to spans ([#2106](https://github.com/getsentry/sentry-dart/pull/2106))
 - Support WebAssembly compilation (dart2wasm) ([#2113](https://github.com/getsentry/sentry-dart/pull/2113))
+- Add memory usage to contexts ([#2133](https://github.com/getsentry/sentry-dart/pull/2133))
```
Comment (buenaflor) [inline CHANGELOG.md]: new release is out, this is now in a wrong section

**Review by buenaflor — CHANGES_REQUESTED** [review]
one last fix pls

File: CHANGELOG.md (hunk 7 above)
Comment (denrase) [reply CHANGELOG.md]: ah sry, one sec

**Review by buenaflor — APPROVED** [review]

**Pull request merged**